src/core/vector/qgsvectorlayerprofilegenerator.cpp

```diff
@@ -802,423 +814,451 @@ bool QgsVectorLayerProfileGenerator::generateProfileForPoints()
+    distanceVHeightRing->append( reversedDistanceVHeightExtrusion.get() );
+    distanceVHeightRing->close();
+    resultFeature.crossSectionGeometry = QgsGeometry( new QgsPolygon( distanceVHeightRing.release() ) );
   }
-  return true;
+  else
+  {
+    resultFeature.geometry = QgsGeometry( new QgsLineString( newX, newY, newZ ) ) ;
+    resultFeature.crossSectionGeometry = QgsGeometry( new QgsLineString( newDistance, newZ ) );
+  }
+
+  mResults->features[resultFeature.featureId].append( resultFeature );
 }
 
 bool QgsVectorLayerProfileGenerator::generateProfileForLines()
 {
   // get features from layer
   QgsFeatureRequest request;
   request.setDestinationCrs( mTargetCrs, mTransformContext );
-  request.setFilterRect( mProfileCurve->boundingBox() );
+  request.setFilterRect( mProfileBox->boundingBox() );
   request.setSubsetOfAttributes( mDataDefinedProperties.referencedFields( mExpressionContext ), mFields );
   request.setFeedback( mFeedback.get() );
 
-  auto processCurve = [this]( const QgsFeature & feature, const QgsCurve * curve )
+  auto processCurve = [this]( const QgsFeature & feature, const QgsCurve * featGeomPart )
   {
     QString error;
-    std::unique_ptr< QgsAbstractGeometry > intersection( mProfileCurveEngine->intersection( curve, &error ) );
+    std::unique_ptr< QgsAbstractGeometry > intersection( mProfileBoxEngine->intersection( featGeomPart, &error ) );
     if ( !intersection )
       return;
 
     if ( mFeedback->isCanceled() )
       return;
 
-    QgsGeos curveGeos( curve );
-    curveGeos.prepareGeometry();
-
-    if ( mFeedback->isCanceled() )
-      return;
 
-    for ( auto it = intersection->const_parts_begin(); it != intersection->const_parts_end(); ++it )
+    // Intersection is empty : GEOS issue for vertical intersection : use feature geometry as intersection
+    if ( intersection->isEmpty() )
     {
-      if ( mFeedback->isCanceled() )
-        return;
+      std::string wkt_std = intersection->asWkt().toStdString();
+      intersection.reset( feature.geometry().constGet()->clone() );
+    }
+
+    QgsGeos featGeomPartGeos( featGeomPart );
+    featGeomPartGeos.prepareGeometry();
 
+    for ( auto it = intersection->const_parts_begin();
+          !mFeedback->isCanceled() && it != intersection->const_parts_end();
+          ++it )
+    {
       if ( const QgsPoint *intersectionPoint = qgsgeometry_cast< const QgsPoint * >( *it ) )
       {
         // unfortunately we need to do some work to interpolate the z value for the line -- GEOS doesn't give us this
-        const double distance = curveGeos.lineLocatePoint( *intersectionPoint, &error );
-        std::unique_ptr< QgsPoint > interpolatedPoint( curve->interpolatePoint( distance ) );
-
-        const double offset = mDataDefinedProperties.valueAsDouble( QgsMapLayerElevationProperties::ZOffset, mExpressionContext, mOffset );
+        QString error;
```

Comment:
for processIntersectionPoint you've removed this same geos workaround -- can you confirm that it ONLY needs to be applied to line results and that geos always gives interpolated z for point results?

Reply:
Open question